### PR TITLE
🐛 `aiida_localhost`: suffix label with xdist-worker id

### LINKED ISSUE #7347: `aiida_profile_clean` causes flaky test failures under pytest-xdist
Might be slop but I am just documenting all flaky CI behavior. This one does not have an easy fix

https://github.com/aiidateam/aiida-core/actions/runs/25106933128/job/73570476035?pr=7284

```
_______________________ ERROR at setup of test_get_by_id _______________________
[gw0] linux -- Python 3.10.20 /home/runner/work/aiida-core/aiida-core/.venv/bin/python3
src/aiida/tools/pytest_fixtures/orm.py:100: in factory
    computer = Computer.collection.get(
src/aiida/orm/entities.py:143: in get
    return res.one()[0]
src/aiida/orm/querybuilder.py:1152: in one
    raise NotExistent('No result was found')
E   aiida.common.exceptions.NotExistent: No result was found

During handling of the above exception, another exception occurred:
.venv/lib/python3.10/site-packages/sqlalchemy/engine/base.py:1965: in _exec_single_context
    self.dialect.do_execute(
.venv/lib/python3.10/site-packages/sqlalchemy/engine/default.py:921: in do_execute
    cursor.execute(statement, parameters)
E   sqlite3.IntegrityError: UNIQUE constraint failed: db_dbcomputer.label

The above exception was the direct cause of the following exception:
src/aiida/storage/psql_dos/orm/utils.py:119: in save
    self.session.commit()
.venv/lib/python3.10/site-packages/sqlalchemy/orm/session.py:1923: in commit
    trans.commit(_to_root=True)
<string>:2: in commit
    ???
.venv/lib/python3.10/site-packages/sqlalchemy/orm/state_changes.py:139: in _go
    ret_value = fn(self, *arg, **kw)
.venv/lib/python3.10/site-packages/sqlalchemy/orm/session.py:1239: in commit
    self._prepare_impl()
<string>:2: in _prepare_impl
    ???
.venv/lib/python3.10/site-packages/sqlalchemy/orm/state_changes.py:139: in _go
    ret_value = fn(self, *arg, **kw)
.venv/lib/python3.10/site-packages/sqlalchemy/orm/session.py:1214: in _prepare_impl
    self.session.flush()
.venv/lib/python3.10/site-packages/sqlalchemy/orm/session.py:4179: in flush
    self._flush(objects)
.venv/lib/python3.10/site-packages/sqlalchemy/orm/session.py:4314: in _flush
    with util.safe_reraise():
.venv/lib/python3.10/site-packages/sqlalchemy/util/langhelpers.py:147: in __exit__
    raise exc_value.with_traceback(exc_tb)
.venv/lib/python3.10/site-packages/sqlalchemy/orm/session.py:4275: in _flush
    flush_context.execute()
.venv/lib/python3.10/site-packages/sqlalchemy/orm/unitofwork.py:466: in execute
    rec.execute(self)
.venv/lib/python3.10/site-packages/sqlalchemy/orm/unitofwork.py:642: in execute
    util.preloaded.orm_persistence.save_obj(
.venv/lib/python3.10/site-packages/sqlalchemy/orm/persistence.py:93: in save_obj
    _emit_insert_statements(
.venv/lib/python3.10/site-packages/sqlalchemy/orm/persistence.py:1226: in _emit_insert_statements
    result = connection.execute(
.venv/lib/python3.10/site-packages/sqlalchemy/engine/base.py:1412: in execute
    return meth(
.venv/lib/python3.10/site-packages/sqlalchemy/sql/elements.py:515: in _execute_on_connection
    return connection._execute_clauseelement(
.venv/lib/python3.10/site-packages/sqlalchemy/engine/base.py:1635: in _execute_clauseelement
    ret = self._execute_context(
.venv/lib/python3.10/site-packages/sqlalchemy/engine/base.py:1844: in _execute_context
    return self._exec_single_context(
.venv/lib/python3.10/site-packages/sqlalchemy/engine/base.py:1984: in _exec_single_context
    self._handle_dbapi_exception(
.venv/lib/python3.10/site-packages/sqlalchemy/engine/base.py:2339: in _handle_dbapi_exception
    raise sqlalchemy_exception.with_traceback(exc_info[2]) from e
.venv/lib/python3.10/site-packages/sqlalchemy/engine/base.py:1965: in _exec_single_context
    self.dialect.do_execute(
.venv/lib/python3.10/site-packages/sqlalchemy/engine/default.py:921: in do_execute
    cursor.execute(statement, parameters)
E   sqlalchemy.exc.IntegrityError: (sqlite3.IntegrityError) UNIQUE constraint failed: db_dbcomputer.label
E   [SQL: INSERT INTO db_dbcomputer (uuid, label, hostname, description, scheduler_type, transport_type, metadata) VALUES (?, ?, ?, ?, ?, ?, ?)]
E   [parameters: ('aa6d37d0-edd5-4661-a4ed-97e5bd90271a', 'localhost', 'localhost', '', 'core.direct', 'core.local', '{"workdir": "/tmp/pytest-of-runner/pytest-0/popen-gw0/test_get_by_id0"}')]
E   (Background on this error at: https://sqlalche.me/e/20/gkpj)
```

and here
https://github.com/aiidateam/aiida-core/actions/runs/25108262868/job/73586466642?pr=7284
```
...................[gw0] node down: Not properly terminated
F
_____________________ tests/storage/psql_dos/test_query.py _____________________
[gw0] linux -- Python 3.10.20 /home/runner/work/aiida-core/aiida-core/.venv/bin/python3
worker 'gw0' crashed while running 'tests/storage/psql_dos/test_query.py::test_qb_ordering_limits_offsets_sqla'

replacing crashed worker gw0
```

and also the REST API error might be related to this
https://github.com/aiidateam/aiida-core/actions/runs/25117114340/job/73607205380
```
❯ Next this one F                                                                                                                                                                               
_________________________ tests/restapi/test_routes.py _________________________                                                                                                                
[gw0] linux -- Python 3.14.4 /home/runner/work/aiida-core/aiida-core/.venv/bin/python3                                                                                                          
worker 'gw0' crashed while running 'tests/restapi/test_routes.py::TestRestApi::test_computers_orderby_schedulertype_desc'                                                                       
                                                                                                                                                                                                
replacing crashed worker gw0
```

> init_profile is an autouse=True fixture that uses aiida_profile_clean — every test method in TestRestApi triggers reset_storage() on the shared database. Under xdist, this
  crashes workers.


## Summary

Several tests fail intermittently under `pytest-xdist` due to a race condition when `aiida_profile_clean` resets the shared database while other workers are actively using it. This manifests as both `IntegrityError` failures and outright worker crashes.

Observed in: https://github.com/aiidateam/aiida-core/actions/runs/25106933128/job/73570476092?pr=7284

## Failure Category 1: UNIQUE constraint violation (7 tests)

**Affected tests** (all in `tests/cmdline/params/types/test_code.py`):
- `test_get_by_id`
- `test_get_by_uuid`
- `test_get_by_label`
- `test_get_by_fullname`
- `test_ambiguous_label_pk`
- `test_ambiguous_label_uuid`
- `test_entry_point_validation`

**Affected jobs:** `test minimum reqs (3.10, sqlite)`, `tests-presto`

**Error:**

```
src/aiida/tools/pytest_fixtures/orm.py:100: in factory
    computer = Computer.collection.get(
E   aiida.common.exceptions.NotExistent: No result was found

During handling of the above exception, another exception occurred:

E   aiida.common.exceptions.IntegrityError: (sqlite3.IntegrityError) UNIQUE constraint failed: db_dbcomputer.label
E   [SQL: INSERT INTO db_dbcomputer ... VALUES (..., 'localhost', ...)]
```

**Root cause:**

The `aiida_profile` fixture is session-scoped, so all xdist workers share the same database. `test_shell_complete` (line 127) uses `aiida_profile_clean`, which calls `reset_storage()` and wipes the entire database. This creates a race condition in the `aiida_computer` fixture (`src/aiida/tools/pytest_fixtures/orm.py:100-111`):

1. Worker A resets the database via `aiida_profile_clean`
2. Worker A calls `Computer.collection.get(label='localhost', ...)` -> `NotExistent`
3. Worker B (or another fixture in the same worker) also calls `.get(...)` -> `NotExistent`
4. Worker A creates and stores the computer -> succeeds
5. Worker B tries to create the same computer -> `IntegrityError` on the label unique constraint

## Failure Category 2: Worker crashes (5 tests)

In the same CI run, `gw0` crashed ("node down: Not properly terminated") during five different tests across four jobs, with no Python traceback:

| Job | Test running when worker crashed |
|-----|----------------------------------|
| `tests-presto` | `tests/tools/archive/orm/test_comments.py::test_exclude_comments_flag` |
| `tests (3.14, psql, rmq)` | `tests/restapi/test_routes.py::TestRestApi::test_computers_orderby_schedulertype_desc` |
| `tests (3.10, psql, zmq)` | `tests/storage/psql_dos/test_backend.py::test_get_info` |
| `tests (3.14, psql, zmq)` | `tests/orm/implementation/test_nodes.py::TestBackendNode::test_clear_attributes` |
| `tests (3.10, psql, rmq)` | `tests/restapi/test_identifiers.py::test_full_type_unregistered[WorkFunctionNode]` |

These are likely caused by the same `reset_storage()` race corrupting database state mid-operation in another worker, leading to segfaults or fatal errors in the storage/ORM layer.

## Possible fixes

1. **Give each xdist worker its own profile/database** -- eliminates all shared-state races but increases resource usage.
2. **Avoid `aiida_profile_clean` in xdist runs** -- use unique labels/data per test instead of wiping the database.
3. **Handle `IntegrityError` in `aiida_computer` fixture** -- catch the error and retry with a `.get()` call:
   ```python
   try:
       computer = Computer(label=label, ...).store()
   except IntegrityError:
       computer = Computer.collection.get(label=label)
   ```
   This only addresses Category 1 and doesn't fix the worker crashes.

Option 1 is the most robust since it eliminates the root cause for both failure categories.

### PR DESCRIPTION
**Pull request by GeigerJ2**
commit msg:
```
🐛 `aiida_localhost`: suffix label with xdist-worker id

Fixes the `UNIQUE constraint failed: db_dbcomputer.label` flake on
`tests-presto` and friends (#7347, Category 1). The `aiida_localhost`
fixture's row and a literal-`'localhost'` row created by another code
path (notably `verdi presto`) ended up in the same profile, and one
of two adjacent statements in the fixture's get-or-create — the four-
field `Computer.collection.get` followed by `Computer(...).store()` —
saw inconsistent state and tripped the single-column UNIQUE on
`Computer.label`. Tracing didn't pin the exact within-worker mechanism
(likely a SQLAlchemy session-cache stale read after `aiida_profile_-
clean` reset state); xdist surfaces it more often by widening the test
ordering space.

Fix: suffix the fixture's label with the pytest-xdist worker id
(`localhost-gw0`, ..., or `localhost-master` outside xdist). The
fixture row and any literal-label row coexist under different names,
so UNIQUE cannot fire regardless of the timing or cache state. The
literal `'localhost'` label remains the production contract for
`verdi presto` and is untouched. `'master'` matches xdist's own
`worker_id` fixture convention.

Eight tests asserted against the literal `'localhost'` form and broke;
they are parameterized on `aiida_localhost.label` so the comparison is
exact and stable across workers. `test_code.py`'s
`_normalize_code_show_output` strips the worker-id suffix from the
fixture's label before the hostname substitution (the suffixed label
contains the hostname as a substring), so the four affected golden
files keep the user-facing literal `localhost`. The three graphviz
tests in `test_graph.py` build their expected strings off
`self.computer.label`.

Audit other test sites that bypassed the fixture and created or loaded
a literal-`'localhost'` Computer directly, switching them to depend on
`aiida_localhost`: `tests/calculations/test_stash.py`,
`tests/orm/nodes/data/test_remote.py`'s factory, and the `TestNode`
class in `tests/orm/nodes/test_node.py` (`setup_method` becomes a
pytest autouse fixture). Literal-`'localhost'` sites in
`test_presto.py` and the fixture's own non-collision regression test
are intentional. Production code paths that own the literal-
`'localhost'` contract (`verdi presto`, `prepare_localhost`,
`stash`/`unstash`) are deliberately untouched.

Drive-by: document `aiida_computer`'s `IntegrityError` recovery —
why catch-and-retry-`get()` rather than loop-with-backoff, why a
defensive `NotExistent` fallback is kept, and why a user-space lock
or `INSERT ... ON CONFLICT` would not be a better fit here.

Addresses Category 1 of #7347 only. Category 2 (worker crashes from
`reset_storage()` mid-test) needs per-worker profile isolation, not a
label fix, and is left for follow-up.
```

### CODE REVIEW

File: tests/orm/nodes/test_node.py [hunk 1]
```diff
@@ -17,24 +17,18 @@
...
 from aiida.orm.utils.links import LinkTriple
 
 
 class TestNode:
     """Tests for generic node functionality."""
 
-    def setup_method(self):
+    @pytest.fixture(autouse=True)
+    def _setup(self, aiida_localhost):
         """Setup for methods."""
```
Comment (GeigerJ2) [inline tests/orm/nodes/test_node.py]: Switched from a unittest-style `setup_method` to a pytest autouse fixture because we need to inject `aiida_localhost`. `setup_method` can't receive fixtures as parameters, and the previous body was creating a literal-`'localhost'` `Computer` directly via `Computer.collection.get_or_create(...)`, which is exactly the kind of fixture-bypass the rest of this PR is auditing away. With the worker-suffixed `aiida_localhost`, that direct creation would leave a stray literal-`'localhost'` row alongside the fixture's `localhost-{worker_id}` row.

Renamed to `_setup` (rather than keeping `setup_method`) because `setup_method` is pytest's xunit-style class setup hook — it's invoked directly with signature `(self, method)`. If we kept that name and added `@pytest.fixture(autouse=True)`, pytest's xunit dispatch would still try to call it as `setup_method(self, method)`, which mismatches our `(self, aiida_localhost)` signature. The leading underscore is purely a convention to flag it as an internal autouse fixture, not meant to be referenced by name.

File: tests/calculations/test_stash.py [hunk 2]
```diff
@@ -109,7 +109,7 @@ def test_code_vs_stash_mode_conflict(stash_mode, fixture_sandbox, aiida_localhos
     code = orm.InstalledCode(
         label='dummy_code',
         default_calc_job_plugin='core.stash',
-        computer=orm.load_computer(label='localhost'),
+        computer=aiida_localhost,
```
Comment (GeigerJ2) [inline tests/calculations/test_stash.py:112]: As a note here, also relevant to the other changes to tests here:
The audit was deliberately scoped to test code only. Production sites that create or load a literal-`'localhost'` Computer are intentionally untouched: `src/aiida/cmdline/commands/cmd_presto.py`, `src/aiida/cmdline/commands/cmd_devel.py` (`prepare_localhost`), `src/aiida/calculations/stash.py`, and `src/aiida/calculations/unstash.py` all stay as-is. Those define the user-facing contract that "the localhost computer's label is `'localhost'`", which is what users see after running `verdi presto`. Routing them through `aiida_localhost` would either change that contract or require a fixture-aware code path in production, neither of which we want. The fix here lives entirely in the test fixture so production behaviour is unchanged.

(Side note unrelated to this PR: `cmd_devel.py:271` has a stale docstring referencing `aiida.engine.launch_shell_job`, which doesn't exist in the codebase — `prepare_localhost` is the actual function. Worth a one-line follow-up but out of scope here.)

File: src/aiida/tools/pytest_fixtures/orm.py [hunk 3]
```diff
@@ -114,10 +123,23 @@ def _build() -> 'Computer':
             try:
                 computer.store()
             except IntegrityError:
```
Comment (GeigerJ2) [inline src/aiida/tools/pytest_fixtures/orm.py:125]: This is still literal Claude output that I have not verified, but it seems sensible:

While putting this fix together I traced the test fixture wiring end-to-end to verify the framing in #7347 (and in @agoscinski's #7349). Two things worth knowing:

1. **Workers don't share databases in the test setup.** `aiida_config` uses `tmp_path_factory.mktemp(secrets.token_hex(16))`; under xdist each worker has its own `basetemp` (`<basetemp>/popen-<worker_id>`, set by `xdist/workermanage.py:339-341`) and its own per-process random state, so each worker's SQLite file lives at a unique path. For PostgreSQL, `postgres_cluster` is session-scoped and uses `pgtest.PGTest()`, which spawns a fresh ephemeral cluster per worker on a random port. The CI's `services: postgres` is not what tests connect to — `pgtest` spawns its own.

2. **So the "another xdist worker resets the shared DB" mechanism in #7347's body and in #7349's commit message can't be the literal cause** of the `IntegrityError` we observe. The trace in #7347 shows the existing row's scheduler/transport matched the fixture's defaults exactly, so the four-field `Computer.collection.get` should have hit it but raised `NotExistent` anyway. Within a single sequential worker process, the only mechanism that fits is a SQLAlchemy session-cache stale read across two adjacent statements (or some weird in-flight transaction state). The label suffix in this PR sidesteps the question entirely by making the fixture's row impossible to collide with.

I kept the `IntegrityError` catch and its `NotExistent` fallback (originally added by @agoscinski in #7349) as defensive belt-and-braces — the suffix should make this branch unreachable from `aiida_localhost`, but direct callers of the lower-level `aiida_computer` factory could still hit it. Comment in `orm.py` documents the no-retry-loop reasoning and why a lock or `INSERT ... ON CONFLICT` would not be a better fit here.

This PR addresses Category 1 of #7347 only (UNIQUE collisions). Category 2 (worker crashes from `reset_storage()` mid-test) is a separate concern that needs per-worker profile isolation rather than a label fix; left for follow-up.

**Review by edan-bainglass — APPROVED** [review]
Thanks @GeigerJ2. Tested this on #6990, which had failed a few tests with `IntegrityError`. Those tests now pass. As stated elsewhere, due to the flaky nature of these tests, we can't with absolute certainty say that this fixed it, but at least from what I can see, the solution here is quite sensible and in theory should resolve the problem consistently.

One comment is perhaps before merging, have a pass over the added inline comments and docs. I found them a bit convoluted. Great if we can simplify the language, or at least try to aim them at our future selves (or AiiDA's future developers). This would be helpful 🙏

Of course, also triple-check anything authored by your agent if you haven't already.

**Comment by agoscinski** [comment]
Skimmed through it. Really complex issue. Is it due to some SQLAlchemy cache that is shared between different xdist workers? I am missing in this long explanations the resource that the xdist workers shared. I though we are acting on different databases for each worker so this should not happen? And if we are not acting on different databases for each worker, wouldn't the easier fix to just fix this on the database level and not on the computer level, so it cannot occur for anything else?

I am fine to remove the previous fix, it was only a tmp fix because we did not know the exact cause. If the reason for the bug is that we actually share the database between workers then this PR should be the real fix and the previous tmp fix can be removed. If we share the database between workers, we should fix this instead. If it is a weird SQLAlchemy shares a cache creating a shared state behind the scene without letting us know, then I am fine with the currently implemented fix.

**Comment by GeigerJ2** [comment]
@agoscinski thanks for pushing on this. I (Claude ^^) went deep into the `SQLAlchemy` and `pytest-xdist` sources to try to actually nail the mechanism, and the honest answer is I (Claude ^^) can't pin it down from static analysis alone. Let me lay out what I'm now confident about vs. what I had to give up on.

tl;dr: workers don't share databases (see below), so the cause is somewhere within a single worker process. My best guess is `SQLite`/`fsync` timing on busy CI runners or a session-state edge case in the `_clear()` + `reset_profile()` + lazy-reinit boundary, but static analysis didn't pin it. The label-suffix avoids the question rather than answering it, but I think that's acceptable because a deterministic DB-level fix wouldn't give us materially more than what we already have.

The chain, end to end:

- `pytest-xdist`'s `workermanage.py:337-341` launches each worker with `basetemp = <main_basetemp>/popen-<gateway.id>` (so `popen-gw0`, `popen-gw1`, ...). Each worker is a separate Python process running its own pytest session.
- aiida-core's `tests/conftest.py:279` overrides the upstream `aiida_profile` fixture, building `storage_config` via `config_sqlite_dos()` (`aiida/tools/pytest_fixtures/storage.py:120-121`), which just returns `{'filepath': str(tmp_path_factory.mktemp('test_sqlite_dos_storage'))}`.
- `tmp_path_factory` is itself rooted at the per-worker basetemp set by `xdist`, so each worker's `mktemp(...)` returns a directory under `popen-<gateway.id>/`.
- The session-scoped `aiida_profile` and `aiida_config` fixtures fire once *per worker process* (each worker has its own session), so each worker generates its own profile name and its own storage dir.

I confirmed this empirically — see the snippet at the bottom of this comment. Drop it into `tests/` and run with `-n 4 --db-backend sqlite`. The test body is parametrized 4 times so `xdist` hands one item to each of the 4 workers; each invocation appends a line to `/tmp/aiida_xdist_demo.log` (a file, because `xdist` captures stdout per worker by default, even with `-s`). After the run, `sort /tmp/aiida_xdist_demo.log` shows one row per worker, like so:

| worker | pid    | profile_name (random hex)          | storage filepath                                                        |
|--------|--------|------------------------------------|-------------------------------------------------------------------------|
| `gw0`  | 410608 | `01f3437cc7de7f9903e2a501f8b3a891` | `/tmp/pytest-of-geiger_j/pytest-24/popen-gw0/test_sqlite_dos_storage0` |
| `gw1`  | 410611 | `8ada8cf42f2ebfc71caa7384adfcb310` | `/tmp/pytest-of-geiger_j/pytest-24/popen-gw1/test_sqlite_dos_storage0` |
| `gw2`  | 410617 | `46ad443996ced887666d42b3eb7ff999` | `/tmp/pytest-of-geiger_j/pytest-24/popen-gw2/test_sqlite_dos_storage0` |
| `gw3`  | 410620 | `5da390af9972fd6bb6615b3c52e258d1` | `/tmp/pytest-of-geiger_j/pytest-24/popen-gw3/test_sqlite_dos_storage0` |

Distinct profile names, distinct filepaths, all under their own `popen-gwN/`. For `PostgreSQL` the `postgres_cluster` fixture spawns a fresh `pgtest.PGTest()` per worker on a random port, so the CI's `services: postgres` is not what tests connect to, it's a per-worker ephemeral cluster. The failing trace in #7347 is consistent with this: workdir is `/tmp/pytest-of-runner/pytest-0/popen-gw0/test_get_by_id0`, i.e. `gw0`'s own basetemp. So the "all `xdist` workers share the same database" framing in the issue body is wrong, and the multi-worker TOCTOU mechanism in #7349's commit message can't be the literal cause.

That leaves a within-worker mechanism, and this is where I get stuck. For the trace to make sense (the `get(label='localhost', hostname='localhost', scheduler='core.direct', transport='core.local')` raises `NotExistent`, then the immediately-following `Computer(...).store()` fails `UNIQUE` on `label`), there has to be a row with `label='localhost'` at `INSERT` time but not at `SELECT` time. Within a single sequential pytest worker process there's no concurrency in user-space code. I checked all the literal-`'localhost'` Computer creators in the suite (`test_node.py::TestNode::setup_method` pre-PR, `test_remote.py`'s factory pre-PR, `verdi presto`, `prepare_localhost` in `cmd_devel.py`) and they all use the exact same scheduler/transport defaults, so if the row exists at `INSERT`, the four-field `get()` should also hit it. There's no filter-mismatch path I can find that would explain the inconsistency.

I also went through `SQLAlchemy`'s session model carefully to rule out the obvious cache-state suspects. `Session.close()` doesn't actually remove the session from the `scoped_session` registry (`session.py:2546-2553` is explicit: "the Session itself does not actually have a distinct closed state"), so the close-then-expunge sequence in `_clear()` is fine. `_clear()` does `DELETE FROM <table>` per reflected table, not `DROP TABLE` (`migrator.py:325-346`), so the schema stays intact and `aiida_profile_clean` leaves a clean empty schema. `scoped_session` is thread-local via `ThreadLocalRegistry` (`util/_collections.py:630-645`), but unless something is spawning threads I'm not seeing, that shouldn't matter for these tests. No smoking gun in the session lifecycle.

My best-guess ranking of plausible culprits, none of which I can pin to a specific line of code:

1. **SQLite write-visibility timing under busy CI I/O.** GitHub runners' shared disks have unpredictable `fsync` latencies. A `SELECT` issued just after a `COMMIT` could plausibly miss the row if the `COMMIT` hasn't fully landed in the SQLite reader's snapshot view. We don't set `WAL` mode, so the default rollback-journal isolation is in play with its known edge cases around `BEGIN`/`COMMIT` timing.
2. **Session-state edge case at the `_clear()` → `reset_profile()` → lazy-reinit boundary.** This is where the trace points (the `get()` raises `NotExistent` immediately after an `aiida_profile_clean`-using test ran), and it's also where the most session/engine state is being thrown away and rebuilt. Static analysis doesn't surface an obvious bug, but narrow-window bugs can hide in handoff code like this.
3. **An `xdist` `load`-distribution test ordering specific to CI.** Default `load` is non-deterministic across runs and across runners, so a specific sequence might tickle a state we wouldn't naturally hit otherwise.

The label-suffix is still the right fix at the test-fixture level regardless of which of these is firing. It makes the conflict structurally impossible by construction.

To your "fix it at the DB level so it can't occur for anything else": I think that's a reasonable instinct, but it's worth noting that "anything else" is in practice an empty set here. The only fixture in the suite that uses a literal label is `aiida_localhost` itself; every other call into `aiida_computer*` defaults to a random `uuid.uuid4().hex`, so by construction those can't collide on `UNIQUE(label)`. So the "computer-level fix" objection is mostly hypothetical: nothing else in the suite is at risk in the first place. Of the DB-level options, `INSERT ... ON CONFLICT DO NOTHING RETURNING` would work and is what I'd reach for if I were rewriting this from scratch, but it bypasses `Computer(...).store()`, so we'd lose the AiiDA-side defaults and validation. Removing the `UNIQUE` constraint on `label` is a non-starter (users rely on labels being unique identifiers for `verdi` / `load_computer`). A `filelock` per profile would serialise but is heavy machinery for a test fixture.

The label-suffix is cheaper than all of those: doesn't touch production code, doesn't touch the schema, doesn't introduce dialect-specific SQL, makes the conflict structurally impossible by ensuring the fixture's row and any literal-`'localhost'` row coexist under different names.

On whether to remove #7349: I'd keep it (i.e. *not* revert it in this PR) as defence-in-depth. The `IntegrityError + NotExistent` fallback is dead code for any aiida-core test using `aiida_localhost` (the suffix makes that branch unreachable), but it still catches the case where a direct caller of the lower-level `aiida_computer` factory passes a literal label, e.g. a third-party plugin test doing `aiida_computer_local(label='localhost')` — which can happen in third-party plugin test suites we don't control. Reverting now buys us nothing concrete and re-opens that hole; keeping costs zero (it's already merged and tested in `tests/tools/pytest_fixtures/test_orm.py`).

<details>
<summary><code>test_xdist_isolation_demo.py</code> — drop into <code>tests/</code> and run with <code>-n 4 --db-backend sqlite</code></summary>

```python
"""Show that pytest-xdist workers each get their own AiiDA profile and SQLite DB.

Drop into tests/ (so the AiiDA pytest plugin is active) and run:

    .venv/bin/python -m pytest tests/test_xdist_isolation_demo.py \
        -n 4 -p no:cacheprovider --db-backend sqlite

Each test invocation appends a line to /tmp/aiida_xdist_demo.log (we write to a
file because xdist captures stdout per worker by default).

After the run, eyeball with:

    sort /tmp/aiida_xdist_demo.log

You should see one line per worker_id (gw0, gw1, ...), each with a distinct
profile_name and a distinct filepath. If you see all workers reporting the
same path, the isolation argument is wrong.
"""

from __future__ import annotations

import os
from pathlib import Path

import pytest

LOG_PATH = Path('/tmp/aiida_xdist_demo.log')


@pytest.fixture(scope='session', autouse=True)
def _ensure_log_exists():
    LOG_PATH.touch(exist_ok=True)
    yield


@pytest.mark.parametrize('iteration', range(4))
def test_per_worker_storage_isolation(aiida_profile, worker_id, iteration):
    """Log this worker's profile + storage path. Different workers should differ."""
    line = (
        f'[{worker_id} pid={os.getpid():>6}] '
        f'profile_name={aiida_profile.name} '
        f'storage_backend={aiida_profile.storage_backend} '
        f'filepath={aiida_profile.storage_config.get("filepath", "<n/a>")}\n'
    )
    with LOG_PATH.open('a') as f:
        f.write(line)

    if aiida_profile.storage_backend == 'core.sqlite_dos':
        storage_dir = Path(aiida_profile.storage_config['filepath'])
        assert storage_dir.is_dir(), f'storage dir missing for {worker_id}: {storage_dir}'
```

</details>
